Unfold measures when using PLE

It will take some debugging to figure out why this workaround was needed in liquidhaskell:
https://github.com/ucsd-progsys/liquidhaskell/pull/1874

I'm guessing that liquid-fixpoint should unfold measures with PLE, but please comment if we should consider other alternatives.

The following change undoes the effect of 1874 on `tests/pos/T1874.hs`
```
diff --git a/src/Language/Haskell/Liquid/Bare.hs b/src/Language/Haskell/Liquid/Bare.hs
index 38e14c564..3d2a0c016 100644
--- a/src/Language/Haskell/Liquid/Bare.hs
+++ b/src/Language/Haskell/Liquid/Bare.hs
@@ -396,7 +396,7 @@ makeLiftedSpec0 :: Config -> GhcSrc -> F.TCEmb Ghc.TyCon -> LogicMap -> Ms.BareS
                 -> Ms.BareSpec
 makeLiftedSpec0 cfg src embs lmap mySpec = mempty
   { Ms.ealiases  = lmapEAlias . snd <$> Bare.makeHaskellInlines (typeclass cfg) src embs lmap mySpec 
-  , Ms.reflects  = Ms.reflects mySpec <> (if reflection cfg then Ms.hmeas mySpec else mempty)
+  , Ms.reflects  = Ms.reflects mySpec -- <> (if reflection cfg then Ms.hmeas mySpec else mempty)
   , Ms.dataDecls = Bare.makeHaskellDataDecls cfg name mySpec tcs  
   , Ms.embeds    = Ms.embeds mySpec
   -- We do want 'embeds' to survive and to be present into the final 'LiftedSpec'. The
```
This has LH fail with
```
**** LIQUID: UNSAFE ************************************************************

T1874.hs:68:15: error:
    Liquid Type Mismatch
    .
    The inferred type
      VV : T1874.Exp
    .
    is not a subtype of the required type
      VV : {VV : T1874.Exp | T1874.exprType e0 == T1874.exprType VV}
    .
    in the context
      e0 : T1874.Exp
    Constraint id 21
   |
68 |               Just e11 -- ? typeAppLam_prop ty e11 e2
   |               ^^^^^^^^
```
when I look at the equations produced by PLE in the constraint 21 I get surprisingly:
```
$ cat tests/pos/.liquid/T1874.hs.fq.ple
...
constraint id 21:
  && [((T1874.isFunTy T1874.TInt) = false);
      ((is$36$GHC.Types.$58$ GHC.Types.$91$$93$) = false);
      ((is$36$GHC.Types.$91$$93$ GHC.Types.$91$$93$) = true);
      ((is$T1874.TFun T1874.TInt) = false);
      ((is$T1874.TInt T1874.TInt) = true);
      ((isJust GHC.Maybe.Nothing) = false);
      ((isJust lq_anf$##7205759403792803251##d2vF) = false);
      ((isJust lq_anf$##7205759403792803253##d2vH) = false);
      ((len GHC.Types.$91$$93$) = 0)]
...
```
Which are too little equations, in particular, there is no sign of the measure function `exprType`, though other equalities from measures are there.

If we instead run the test with the unmodified LH, the module is safe and more equations are output:
```
$ cat tests/pos/.liquid/T1874.hs.fq.ple
...
constraint id 21:
  && [((T1874.exprType VV##F##21) =
         (if (is$T1874.Var VV##F##21) then
            (lqdc##$select##T1874.Var##2 VV##F##21)
          else
            (if (is$T1874.Lam VV##F##21) then
               (T1874.TFun (lqdc##$select##T1874.Lam##2 VV##F##21) (T1874.exprType (lqdc##$select##T1874.Lam##1 VV##F##21)))
             else
               (T1874.funResTy (T1874.exprType (T1874.e1 VV##F##21))))));
      ((T1874.exprType e0##aCL) =
         (T1874.funResTy (T1874.exprType (T1874.e1 e0##aCL))));
      ((T1874.exprType e0##aCL) =
         (if (is$T1874.Var (lqdc##$select##T1874.Lam##1 (T1874.e1 e0##aCL))) then
            (lqdc##$select##T1874.Var##2 (lqdc##$select##T1874.Lam##1 (T1874.e1 e0##aCL)))
          else
            (if (is$T1874.Lam (lqdc##$select##T1874.Lam##1 (T1874.e1 e0##aCL))) then
               (T1874.TFun (lqdc##$select##T1874.Lam##2 (lqdc##$select##T1874.Lam##1 (T1874.e1 e0##aCL))) (T1874.exprType (lqdc##$select##T1874.Lam##1 (lqdc##$select##T1874.Lam##1 (T1874.e1 e0##aCL)))))
             else
               (T1874.funResTy (T1874.exprType (T1874.e1 (lqdc##$select##T1874.Lam##1 (T1874.e1 e0##aCL))))))));
      ((T1874.exprType e1##aCM) =
         (T1874.TFun (lqdc##$select##T1874.Lam##2 e1##aCM) (T1874.exprType (lqdc##$select##T1874.Lam##1 e1##aCM))));
    ...
...
```

The fix in ucsd-progsys/liquidhaskell#1874 is putting `exprType` both as a measure and as a reflected function, but even if it were only a measure, PLE should pick it up.

cc @nikivazou

--- a/src/Language/Fixpoint/Solver/PLE.hs
+++ b/src/Language/Fixpoint/Solver/PLE.hs
@@ -401,7 +401,15 @@ makeCandidates :: Knowledge -> ICtx -> Expr -> [Expr]
 makeCandidates γ ctx expr 
   = mytracepp ("\n" ++ show (length cands) ++ " New Candidates") cands
   where 
-    cands = filter (\e -> isRedex γ e && not (e `S.member` icSolved ctx)) (notGuardedApps expr)
+    cands =
+      filter (\e -> isRedex γ e && not (e `S.member` icSolved ctx)) (notGuardedApps expr) ++
+      filter (\e -> hasConstructors γ e && not (e `S.member` icSolved ctx)) (largestApps expr)
+
+    -- Constructor occurrences need to be considered as candidadates since
+    -- they identify relevant measure equations. The function 'rewrite'
+    -- introduces these equations.
+    hasConstructors :: Knowledge -> Expr -> Bool
+    hasConstructors γ e =  not $ S.null $ S.intersection (exprSymbolsSet e) (knDCs γ)
 
 isRedex :: Knowledge -> Expr -> Bool 
 isRedex γ e = isGoodApp γ e || isIte e 
@@ -507,6 +515,35 @@ notGuardedApps = flip go []
       PExist _ _ -> acc
       PGrad{} -> acc
 
+-- | @largestApps e@ yields all the largest subexpressions that are
+-- applications not under an if-then-else, lambda abstraction, type abstraction,
+-- type application, or quantifier.
+largestApps :: Expr -> [Expr]
+largestApps = flip go []
+  where
+    go e0 acc = case e0 of
+      EApp _ _ -> e0 : acc
+      PAnd es -> foldr go acc es
+      POr es -> foldr go acc es
+      PAtom _ e1 e2 -> go e1 $ go e2 acc
+      PIff e1 e2 -> go e1 $ go e2 acc
+      PImp e1 e2 -> go e1 $ go e2 acc
+      EBin  _ e1 e2 -> go e1 $ go e2 acc
+      PNot e -> go e acc
+      ENeg e -> go e acc
+      EIte b _ _ -> go b $ e0 : acc
+      ECoerc _ _ e -> go e acc
+      ECst e _ -> go e acc
+      ESym _ -> acc
+      ECon _ -> acc
+      EVar _ -> e0 : acc
+      ELam _ _ -> acc
+      ETApp _ _ -> acc
+      ETAbs _ _ -> acc
+      PKVar _ _ -> acc
+      PAll _ _ -> acc
+      PExist _ _ -> acc
+      PGrad{} -> acc
 
 
 -- The FuncNormal and RWNormal evaluation strategies are used for REST

--- a/src/Language/Fixpoint/Solver/PLE.hs
+++ b/src/Language/Fixpoint/Solver/PLE.hs
@@ -481,30 +481,31 @@ evalOne γ env ctx _ e | otherwise = do
 -- applications not under an if-then-else, lambda abstraction, type abstraction,
 -- type application, or quantifier.
 notGuardedApps :: Expr -> [Expr]
-notGuardedApps = go 
-  where 
-    go e@(EApp e1 e2)  = [e] ++ go e1 ++ go e2
-    go (PAnd es)       = concatMap go es
-    go (POr es)        = concatMap go es
-    go (PAtom _ e1 e2) = go e1  ++ go e2
-    go (PIff e1 e2)    = go e1  ++ go e2
-    go (PImp e1 e2)    = go e1  ++ go e2 
-    go (EBin  _ e1 e2) = go e1  ++ go e2
-    go (PNot e)        = go e
-    go (ENeg e)        = go e
-    go e@(EIte b _ _)  = go b ++ [e] -- ++ go e1 ++ go e2  
-    go (ECoerc _ _ e)  = go e 
-    go (ECst e _)      = go e 
-    go (ESym _)        = []
-    go (ECon _)        = []
-    go (EVar _)        = []
-    go (ELam _ _)      = []
-    go (ETApp _ _)     = []
-    go (ETAbs _ _)     = []
-    go (PKVar _ _)     = []
-    go (PAll _ _)      = []
-    go (PExist _ _)    = []
-    go (PGrad{})       = []
+notGuardedApps = flip go []
+  where
+    go e0 acc = case e0 of
+      EApp e1 e2 -> e0 : go e1 (go e2 acc)
+      PAnd es    -> foldr go acc es
+      POr es     -> foldr go acc es
+      PAtom _ e1 e2 -> go e1 $ go e2 acc
+      PIff e1 e2 -> go e1 $ go e2 acc
+      PImp e1 e2 -> go e1 $ go e2 acc
+      EBin  _ e1 e2 -> go e1 $ go e2 acc
+      PNot e -> go e acc
+      ENeg e -> go e acc
+      EIte b _ _ -> go b $ e0 : acc
+      ECoerc _ _ e -> go e acc
+      ECst e _ -> go e acc
+      ESym _ -> acc
+      ECon _ -> acc
+      EVar _ -> acc
+      ELam _ _ -> acc
+      ETApp _ _ -> acc
+      ETAbs _ _ -> acc
+      PKVar _ _ -> acc
+      PAll _ _ -> acc
+      PExist _ _ -> acc
+      PGrad{} -> acc
 
 
 

